site/content/ytt/docs/develop/how-to-modularize.md

```diff
@@ -0,0 +1,472 @@
+
+### Functions
+[Functions](lang-ref-def.md) provide a way to extract common code into a separate fragment or code snippet.
+
+There are two ways to define a function in ytt, a Starlark function, or a YAML fragment function.
+
+Starlark functions make use of a `return` statement. Because of this they can be great for returning a value that must be transformed in some way.
+
+YAML fragment functions differ in that they are YAML structure wrapped in a Starlark function definition. Everything inside the function will be the return value. They can be great when needing to return nested YAML structure, or key and value pairs.
+
+Going back to the previous solution, we can see each `labels` key is duplicated YAML, like `app.kubernetes.io/version: #@ version`. There is also some duplicated string manipulation in the `metadata.name` key.
+```yaml
+#! config.yml
+
+#@ name = "frontend"
+#@ namespace = "default"
+#@ version = "0.1.0"
+#@ replicas = 1
+
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: #@ name + "-deployment"
+  namespace: #@ namespace
+  labels:
+    - app.kubernetes.io/version: #@ version
+    - app.kubernetes.io/name: #@ name
+spec:
+  selector:
+    matchLabels:
+      app: #@ name
+  replicas: #@ replicas
+  template:
+    metadata:
+      labels:
+        - app.kubernetes.io/version: #@ version
+        - app.kubernetes.io/name: #@ name
+    spec:
+      containers:
+        - name: #@ name
+          image: index.docker.io/k14s/image@sha256:6ab29951e0207fde6760f6db227f218f20e875f45b22e8ca0ee06c0c8cab32cd
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: #@ name + "-service"
+  labels:
+    - app.kubernetes.io/version: #@ version
+    - app.kubernetes.io/name: #@ name
+spec:
+  type: ClusterIP
+  ports:
+    - port: 80
+```
+
+Move the duplicated `labels` keys into a YAML fragment function, and move name calculation into a Starlark function.
+
+```yaml
+#! config.yml
+
+#@ name = "frontend"
+#@ namespace = "default"
+#@ version = "0.1.0"
+#@ replicas = 1
+
+#! Starlark function
+#@ def fmt(name, type):
+#@   return "{}-{}".format(name, type)
+#@ end
+
+#! YAML fragment function 
+#@ def labels(name, version):
+app.kubernetes.io/version: #@ version
+app.kubernetes.io/name: #@ name
+#@ end
+
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: #@ fmt(name, "deployment")
+  namespace: #@ namespace
+  labels: #@ labels(name, version)
+spec:
+  replicas: #@ replicas
+  template:
+    spec:
+      containers:
+        - name: #@ name
+          image: index.docker.io/k14s/image@sha256:6ab29951e0207fde6760f6db227f218f20e875f45b22e8ca0ee06c0c8cab32cd
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: #@ fmt(name, "service")
+  labels: #@ labels(name, version)
+spec:
+  type: ClusterIP
+  ports:
+    - port: 80
+```
+Execute this template by running `ytt -f config.yml`.
+Again, the result is identical to our original template, and we can be sure all our repeated sections of code will be consistent.
```

Comment:
> Do you mean to point at the boilerplate YAML?
Do you mean do I mean to reference the original YAML that will be the output? 

No, I don't think its necessary to repeat it here since its rather long and repetitive. What do you think?